src/Kiota.Builder/Refiners/TypeScriptRefiner.cs

```diff
@@ -88,37 +81,421 @@ public override void Refine(CodeNamespace generatedCode)
         AddQueryParameterMapperMethod(
             generatedCode
         );
+        AddModelsInterfaces(generatedCode);
+        ReplaceRequestConfigurationsQueryParamsWithInterfaces(generatedCode);
+    }
+
+    private static void ReplaceRequestConfigurationsQueryParamsWithInterfaces(CodeElement currentElement)
+    {
+        if (currentElement is CodeClass codeClass && (codeClass.IsOfKind(CodeClassKind.QueryParameters) || codeClass.IsOfKind(CodeClassKind.RequestConfiguration)))
+        {
+
+            var targetNS = codeClass.GetImmediateParentOfType<CodeNamespace>();
+            var existing = targetNS.FindChildByName<CodeInterface>(codeClass.Name, false);
+            if (existing != null)
+                return;
+
+            var insertValue = new CodeInterface
+            {
+                Name = codeClass.Name,
+                Kind = codeClass.IsOfKind(CodeClassKind.QueryParameters) ? CodeInterfaceKind.QueryParameters : CodeInterfaceKind.RequestConfiguration
+            };
+            targetNS.RemoveChildElement(codeClass);
+            var codeInterface = targetNS.AddInterface(insertValue).First();
+
+            foreach (var mprop in codeClass.Properties)
+            {
+                codeInterface.AddProperty(mprop);
+            }
+
+            foreach (var codeUsing in codeClass.Usings)
+            {
+                codeInterface.AddUsing(codeUsing);
+            }
+        }
+        CrawlTree(currentElement, x => ReplaceRequestConfigurationsQueryParamsWithInterfaces(x));
+    }
+
+    private static void AddModelsInterfaces(CodeElement generatedCode)
+    {
+        GenerateModelInterfaces(
+           generatedCode,
+           x => $"{x.Name.ToFirstCharacterUpperCase()}Interface".ToFirstCharacterUpperCase()
+       );
+
+       RenameModelInterfacesAndClasses(generatedCode);
+
+    }
+
+    /// <summary>
+    /// Removes the "Interface" suffix temporarily added to model interface names
+    /// Adds the "Impl" suffix to model class names.
+    /// </summary>
+    /// <param name="currentElement"></param>
+    private static void RenameModelInterfacesAndClasses(CodeElement currentElement)
+    {
+        if (currentElement is CodeClass currentClass && currentClass.IsOfKind(CodeClassKind.Model))
+        {
+            currentClass.Name = currentClass.Name + ModelClassSuffix;
+        }
+        if (currentElement is CodeInterface modelInterface && modelInterface.IsOfKind(CodeInterfaceKind.Model))
+        {
+            if (modelInterface.Name.EndsWith("Interface"))
+            {
+                modelInterface.Name = modelInterface.Name.Split("Interface")[0];
+            }
+        }
+
+        CrawlTree(currentElement, x => RenameModelInterfacesAndClasses(x));
+    }
+
+    private static void GenerateModelInterfaces(CodeElement currentElement, Func<CodeClass, string> interfaceNamingCallback)
```

Comment:
instead of duplicating all of that, can't we reuse what we already have in the common refiner? even if it requires adding a few additional parameters

Reply:
There are multiple customizations specific to TypeScript design because of which it is convenient and readable to keep the logic separate. I started with that code in common language refiner and eventually maintained a separate one . There are specifics plugged in during the iteration of the usings, properties and methods to achieve the design 


Some examples are,
- While generating the interfaces they are temporarily named with `interface` suffix and finally this suffix is removed and add the "Impl" suffix to the model classes. 
- Usings are not removed in model class in TS 
- A parameter of model interface type is added in the model constructor.
- The return type of ReplaceType is different. 
- For request config and query parameters the classes are deleted from the namespace once the interfaces are generated.

Reply:
> While generating the interfaces they are temporarily named with interface suffix and finally this suffix is removed and add the "Impl" suffix to the model classes.

Why this temporary name?

> Usings are not removed in model class in TS

Could be driven by a boolean option

> A parameter of model interface type is added in the model constructor.

Could be driven by a boolean option

> The return type of ReplaceType is different.

Can you expand on that?

> For request config and query parameters the classes are deleted from the namespace once the interfaces are generated.

Could be driven by a boolean option

Reply:
The temporary name is because there is a condition that does not allow duplicate element names in the same namespace, that is, within the same DOM element. I appended the "Impl" after the generation so as to avoid missing the update to any references to the class using the class name.

Reply:
but why do you need a temporary name for the interface instead of 1. removing the class, 2. generating the interface, 3. renaming the class, 4. adding back both the interface and the class?

Reply:
> The return type of ReplaceType is different.

https://github.com/microsoft/kiota/blob/c9c6bd88bc89bfa67882b4ec36a1a4e4b0ea1e93/src/Kiota.Builder/Refiners/TypeScriptRefiner.cs#L607

This is different from the function in the `CommonRefiner ReplaceTypeByInterfaceType`.

Reply:
Created a separate issue for this https://github.com/microsoft/kiota/issues/1608.

I am currently focusing on https://github.com/microsoftgraph/msgraph-sdk-javascript/issues/824 and https://github.com/microsoft/kiota/issues/938 and there will updates in the design. 

It would be safer to unify the code after that.